src/high-level/matrix.lisp

```diff
@@ -0,0 +1,288 @@
+;;;; matrix.lisp
+;;;;
+;;;; Author: Cole Scott
+;;;;         Robert Smith
+
+(in-package #:magicl)
+
+(deftype matrix-storage (&optional type)
+  `(simple-array ,type (*)))
+
+(defclass matrix (abstract-tensor)
+  (;; abstract-tensor slots
+   (nrows
+    :initarg :nrows
+    :initform 0
+    :reader nrows
+    :type alexandria:positive-fixnum
+    :documentation "The number of rows in the matrix")
+   (ncols
+    :initarg :ncols
+    :initform 0
+    :reader ncols
+    :type alexandria:positive-fixnum
+    :documentation "The number of columns in the matrix")
+   (size
+    :initarg :size
+    :initform 0
+    :reader size
+    :type (alexandria:positive-fixnum)
+    :documentation "Total number of elements in the matrix")
+   (element-type
+    :initarg :element-type
+    :initform (error "element-type must be specified when creating a matrix instance") ; TODO: much better error messages
+    :reader element-type
+    :type type
+    :documentation "The type of the elements in the matrix")
+   ;; matrix-specific slots
+   (storage
+    :initarg :storage
+    :initform (error "storage must be specified when creating a matrix instance")
+    :reader storage
+    :documentation "Storage of the matrix, typically in a vector in column major order")
+   (order
+    :initarg :order
+    :initform :column-major
+    :reader order
+    :type (member :row-major :column-major)
+    :documentation "Indexing order of storage (:column-major or :row-major)."))
+  (:metaclass abstract-class:abstract-class))
+
+(defun pprint-matrix (stream matrix)
```

Comment:
Could you please add two `&optional` arguments so that we can call this within a `format` control string? See http://www.lispworks.com/documentation/HyperSpec/Body/22_ced.htm